### ISSUE TITLE
Report logical contradictions

### ISSUE THREAD

**Issue by skshetry** [PyCQA/pylint#6076]
### Current problem

```python
class FileSystem:
    @contextmanager
    def open(self, url, mode="r", encoding=None):
        assert mode
        with urllib.request.urlopen(url) as fobj:
  
            if "mode" == "rb":
                yield fobj

            yield io.TextIOWrapper(fobj, encoding=encoding)
```

There was a discussion on adding support for logical tautologies and contradictions in #2051.

### Desired solution

I would expect pylint to raise warning for `"mode" == "rb"` comparisons. Similarly, for other literals (`3 == 4` and `3 != 3`), and between literals of different types `"mode" == 0`.

**Comment by Pierre-Sassoulas** [comment]
Make sense, it could be added to the ``ComparisonChecker`` see original commit for ``comparison-with-itself``: https://github.com/PyCQA/pylint/commit/5f5026627520bd06dae00f9cb87148ba9733d312

**Comment by omarandlorraine** [comment]
Is there a way to evaluate the expression during the runtime of pylint?

The way I think it needs to be done is say, okay the operator is one of `==`, `!=`. `>`, `in`, etc. etc. and the two expressions on either side are constants. Then evaluate the whole thing. This evaluation will yield either `True` or `False`. So your example up there, `if "mode" == "rb":`, is really equivalent to `if False:`. 

Is `if False:` and `if True:` actually worth reporting as well? If so, the check should maybe be "The if-expression is constant" or something like that. The same of course goes for `if 3:`, and `while`, `assert` and maybe others.

What about:

```
def get_size_of_foobar():
    return 0 # TODO: figure this out later

if get_size_of_foobar() > SIZE_OF_CONTAINER:
    print("the foobar is too big for the container")
```

What do you think pylint should do with this?

**Comment by Pierre-Sassoulas** [comment]
astroid can infer value but it's costly (performance) and it can fail, so generally keeping a sane approach to catch the most obvious problem and not have false positives is preferred. i.e. Maybe we should start with comparison between literals for which there could be no inferring mistakes.

**Comment by omarandlorraine** [comment]
In that case, the warning should say something like "this comparison is always False", and "this comparison is always True". Is that correct?

Also which category should this be in: W, R, maybe E?

**Comment by Pierre-Sassoulas** [comment]
Probably ``R`` like  ``comparison-with-itself``: https://github.com/PyCQA/pylint/commit/5f5026627520bd06dae00f9cb87148ba9733d312. ``E`` is reserved for thing that would create an actual error at runtime.
